pkg/datasource/data.go

```diff
@@ -0,0 +1,439 @@
+	return ds, nil
+}
+
+func (ds *dataSource) registerPool() {
+	ds.dPool.New = func() any {
+		d := &data{
+			Payload: make([][]byte, ds.payloadCount),
+		}
+		for i := range d.Payload {
+			d.Payload[i] = make([]byte, 0)
+		}
+		return d
+	}
+}
+
+func (ds *dataSource) Name() string {
+	return ds.name
+}
+
+func (ds *dataSource) Type() Type {
+	return ds.dType
+}
+
+func (ds *dataSource) NewData() Data {
+	return ds.dPool.Get().(Data)
+}
+
+func (ds *dataSource) ByteOrder() binary.ByteOrder {
+	return ds.byteOrder
+}
+
+func resolveNames(id uint32, fields []*field, parentOffset uint32) (string, error) {
+	if id >= uint32(len(fields)) {
+		return "", errors.New("invalid parent id")
+	}
+	out := ""
+	if FieldFlagHasParent.In(fields[id].Flags) {
+		p, err := resolveNames(fields[id].Parent-parentOffset, fields, parentOffset)
+		if err != nil {
+			return "", errors.New("parent not found")
+		}
+		out = p + "."
+	}
+	out += fields[id].Name
+	return out, nil
+}
+
+// AddStaticFields adds a statically sized container for fields to the payload and returns an accessor for the
+// container; if you want to access individual fields, get them from the DataSource directly
+func (ds *dataSource) AddStaticFields(size uint32, fields []StaticField) (FieldAccessor, error) {
+	ds.lock.Lock()
+	defer ds.lock.Unlock()
+
+	idx := ds.payloadCount
+
+	// temporary write to newFields to not write to ds.fields in case of errors
+	newFields := make([]*field, 0, len(fields))
+
+	parentOffset := len(ds.fields)
+	parentFields := make(map[int]struct{})
+	checkParents := make(map[*field]struct{})
+
+	for _, f := range fields {
+		if _, ok := ds.fieldMap[f.FieldName()]; ok {
+			return nil, fmt.Errorf("field %q already exists", f.FieldName())
+		}
+		fieldName := f.FieldName()
+		nf := &field{
+			Name:         fieldName,
+			Index:        uint32(len(ds.fields) + len(newFields)),
+			PayloadIndex: idx,
+			Flags:        FieldFlagStaticMember.Uint32(),
+		}
+		nf.Size = f.FieldSize()
+		nf.Offs = f.FieldOffset()
+		if nf.Offs+nf.Size > size {
+			return nil, fmt.Errorf("field %q exceeds size of container (offs %d, size %d, container size %d)", nf.Name, nf.Offs, nf.Size, size)
+		}
+		if s, ok := f.(TypedField); ok {
+			nf.Kind = s.FieldType()
+		}
+		if tagger, ok := f.(TaggedField); ok {
+			nf.Tags = tagger.FieldTags()
+		}
+		if s, ok := f.(FlaggedField); ok {
+			nf.Flags |= uint32(s.FieldFlags())
+		}
+		if s, ok := f.(AnnotatedField); ok {
+			nf.Annotations = s.FieldAnnotations()
+		}
+		if s, ok := f.(ParentedField); ok {
+			parent := s.FieldParent()
+			if parent >= 0 {
+				nf.Parent = uint32(parent + parentOffset)
+				nf.Flags |= FieldFlagHasParent.Uint32()
+				parentFields[parent] = struct{}{}
+				checkParents[nf] = struct{}{}
+			}
+		}
+		newFields = append(newFields, nf)
+	}
+
+	// Unref parent fields
+	for p := range parentFields {
+		FieldFlagUnreferenced.AddTo(&newFields[p].Flags)
+	}
+
+	// Check whether parent id is valid
```

Comment:
Couldn't this check be made where you add the key to the map `checkParents[nf] = struct{}{}`? (Map is also not needed anymore)

```golang
if parent >= len(fields) {
    return nil, fmt.Errorf("invalid parent for field %q", f.Name)
}
```